rotors_gazebo_plugins/src/gazebo_wind_plugin.cpp

```diff
@@ -145,5 +191,183 @@ void GazeboWindPlugin::OnUpdate(const common::UpdateInfo& _info) {
   wind_speed_pub_.publish(wind_speed_msg);
 }
 
+void GazeboWindPlugin::ReadCustomWindField(std::string& custom_wind_field_path) {
+  std::ifstream fin;
+  fin.open(custom_wind_field_path);
+  if (fin.is_open()) {
+    std::string data_name;
+    float data;
+
+    // Read the line with the variable name
+    while (fin >> data_name) {
+      // Save data on following line into the correct variable
+      if (data_name == "min_x:") {
+        fin >> min_x_;
+      } else if (data_name == "min_y:") {
+        fin >> min_y_;
+      } else if (data_name == "n_x:") {
+        fin >> n_x_;
+      } else if (data_name == "n_y:") {
+        fin >> n_y_;
+      } else if (data_name == "res_x:") {
+        fin >> res_x_;
+      } else if (data_name == "res_y:") {
+        fin >> res_y_;
+      } else if (data_name == "vertical_spacing_factors:") {
+        while (fin >> data) {
+          vertical_spacing_factors_.push_back(data);
+          if (fin.peek() == '\n') break;
+        }
+      } else if (data_name == "bottom_z:") {
+        while (fin >> data) {
+          bottom_z_.push_back(data);
+          if (fin.peek() == '\n') break;
+        }
+      } else if (data_name == "top_z:") {
+        while (fin >> data) {
+          top_z_.push_back(data);
+          if (fin.peek() == '\n') break;
+        }
+      } else if (data_name == "u:") {
+        while (fin >> data) {
+          u_.push_back(data);
+          if (fin.peek() == '\n') break;
+        }
+      } else if (data_name == "v:") {
+        while (fin >> data) {
+          v_.push_back(data);
+          if (fin.peek() == '\n') break;
+        }
+      } else if (data_name == "w:") {
+        while (fin >> data) {
+          w_.push_back(data);
+          if (fin.peek() == '\n') break;
+        }
+      } else {
+        // If invalid data name, read the rest of the invalid line, publish a message and ignore data on next line. Then resume reading.
+        std::string restOfLine;
+        getline(fin, restOfLine);
+        gzerr << "Invalid data name '" << data_name << restOfLine << "' in custom wind field text file. Ignoring data on next line.\n";
+        fin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+      }
+    }
+
+    fin.close();
+
+    ROS_INFO_STREAM("Custom wind field read successfully from text file.");
+  } else {
+    gzerr << "Could not open custom wind field text file.\n";
+  }
+
+}
+
+void GazeboWindPlugin::InterpolateWindVelocity(math::Vector3 link_position, int i_inf, int i_sup, int j_inf, int j_sup, float vertical_spacing_factor_0, float vertical_spacing_factor_1, float vertical_spacing_factor_4, float vertical_spacing_factor_5, math::Vector3& wind_velocity) {
```

Comment:
Can you give a reference to where you got these implementation-ideas of this tri-linear interpolation from (e.g. Wikipedia site, paper,...) ?